### ISSUE TITLE
[Bug]: `rcdefaults` doesn't work? 

### ISSUE THREAD

**Issue by chahak13** [matplotlib/matplotlib#25855]
### Bug summary

This is the `rcdefaults()` function
https://github.com/matplotlib/matplotlib/blob/8faa835c81d9cf12342992521ddb95e9a645bc36/lib/matplotlib/__init__.py#L1068-L1091

What exactly are we trying to do here? According to the docstring, if we're trying to reset the rcParams dictionary and only reset the keys that are not in `STYLE_BLACKLIST` then there are 2 problems I think:
1. `rcParams.clear()` doesn't work. It doesn't clear out the entire dictionary.
2. After the `update` line, `rcParams` will still have keys from `STYLE_BLACKLIST` which, if I were to perform the same operations on a dictionary, shouldn't be the case.
3. While it does make sense to still have the keys from `STYLE_BLACKLIST` but then I'm not sure why `clear()` is used and not just directly update the keys.

Also, sidenote, `popitem()` errors out without a good error message (which is also why clear doesn't actually empty the dict?)

### Code for reproduction

```python
In [3]: mpl.rcParams["webagg.port"]
Out[3]: 8988

In [4]: mpl.rcParams["webagg.port"] = 9000

In [5]: mpl.rcParams["webagg.port"]
Out[5]: 9000

In [6]: mpl.rcdefaults()

In [7]: mpl.rcParams["webagg.port"]
Out[7]: 9000

In [8]: mpl.rcParams.clear()

In [9]: mpl.rcParams["webagg.port"]
Out[9]: 9000
```

### Actual outcome

`clear()` doesn't empty the `rcParams` dictionary.

### Expected outcome

`clear()` should empty the dictionary and after `rcdefaults()`, keys from `STYLE_BLACKLIST` shouldn't be there.

### Additional information

I'm not sure if this is a documentation issue and the behaviour is intended or the behaviour isn't intended.

### Matplotlib Version

Current main

### Installation

None

**Comment by chahak13** [comment]
Seems like @timhoffm  came across this while doing something similar in https://github.com/matplotlib/matplotlib/pull/12577#issuecomment-431620972

**Comment by ksunden** [comment]
`webagg.port` is specifically in the `STYLE_BLACKLIST` which is explicitly excluded by the documentation for `rcdefaults`

The dict inheritance makes it confusing, as we don't actually directly implement `__delitem__` but I am surprised that `clear` _doesn't_ clear out everything.

If I call `dict.clear(mpl.rcParams)` it _does_ clear out everything...

Somehow we are getting `delitem` and `clear` from `MutableMapping` and not actually properly inheriting from `dict` as was likely intended.

**Comment by ksunden** [comment]
For completeness, here is the `STYLE_BLACKLIST`:
{
    'webagg.port_retries',
    'figure.raise_window',
    'docstring.hardcopy',
    'backend',
    'backend_fallback',
    'toolbar',
    'figure.max_open_warning',
    'date.epoch',
    'timezone',
    'webagg.port',
    'webagg.open_in_browser',
    'webagg.address',
    'interactive',
    'tk.window_focus',
    'savefig.directory'
}
```
```python

**Comment by chahak13** [comment]
Yep. I get that the `STYLE_BLACKLIST` is excluded but if `clear()` works as it normally would, then that means that `webagg.port` and any other entry in `STYLE_BLACKLIST` should _not_ be in there.

**Comment by chahak13** [comment]
I think the reason it gets the `MutableMapping` is because our inheritance is defined as `MutableMapping` first and then `dict`.

**Comment by ksunden** [comment]
In general inheriting from `dict` was a mistake (and that is what #25617 is addressing)


If I do `del mpl.rcParams["backend"]` I get `KeyError` (despite the key actually being there)

Yeah, I think that `rcdefaults` is a no-op currently, which is not intended...

We do not test it at all.

Given that it broke and nobody complained... perhaps that is a sign it should be either deprecated completely or reworked entirely as part of #25617... It perhaps gives some latitude to make changes that might otherwise have warranted a difficult to navigate deprecation dance.

With the chainmap implementation (modulo the sub-dict structures) the documented behavior appears to be something like:

```python
non_style = {k: self[k] for k in STYLE_BLACKLIST}
self.maps = [defaults, non_style]
self.maps.new_child()
```

(this is from the perspective of `self` _being_ a `ChainMap` and `defaults` being available, but more or less that is I think what the documented behavior is)

And of course, should get a test...
